packages/common/src/utils/internal/userAppResources.ts

```diff
@@ -0,0 +1,174 @@
+// These Types and Functions can land in rest-js
+// a friendly name for the scope of the resource
+// the API deals with clientIds, which are
+// listed here https://confluencewikidev.esri.com/pages/viewpage.action?pageId=50561461
+// Generally speaking, the UserResourceScope is a downcased version of the
+// clientId. Hub is unique b/c Sites have their own clientId,
+
+import { IRequestOptions, request } from "@esri/arcgis-rest-request";
+import { failSafe } from "../fail-safe";
+import { getObjectSize } from "../getObjectSize";
+
+// Defines the access for the resource
+export type UserResourceAccess =
+  | "userappprivate" // default
+  | "userprivateallapps" // readable by any platform app, private to user
+  | "allorgusersprivateapp" // only if user is public or org
+  | "public"; // only if user is public or org
+
+/**
+ * Add an User-App Resource
+ */
+export interface IAddUserResource {
+  key: string; // aka filename
+  data: Record<string, any>; // json object to store; will be stringified
+  access: UserResourceAccess; // access level
+}
+
+// Returned when we get a list of resources for a user
+export interface IUserResourceInfo {
+  key: string;
+  clientId: string;
+  created: string;
+  size: number;
+  access: UserResourceAccess;
+}
+
+export interface IUserResourceListResponse {
+  total: number;
+  start: number;
+  num: number;
+  nextStart: number;
+  userResources: IUserResourceInfo[];
+}
+
+/**
+ * Set a User-App-Resource
+ * By default this will fetch and merge with an existing resource.
+ * Passing `true` as the last parameter,
+ * @param resource
+ * @param username
+ * @param portalUrl
+ * @param token
+ * @param replace
+ * @returns
+ */
+export async function setUserResource(
+  resource: IAddUserResource,
+  username: string,
+  portalUrl: string,
+  token: string,
+  replace: boolean = false
+): Promise<void> {
+  // Ensure we are below 5MB max size
+  if (getObjectSize(resource.data).megabytes > 4.95) {
+    throw new Error(
+      `User Resource is too large to store. Please ensure it is less than 5MB in size`
+    );
+  }
```

Comment:
This all makes sense. A question though, When setting a resource in this manner can we chunk that resource upload in a manner similar to what we are doing for uploading a content item from the local machine?

Reply:
Platform limit it to 5mb, and it only accepts json - so I'm not sure how we'd chunk it